src/main/flight/imu.c

```diff
@@ -529,3 +526,73 @@ void imuSetHasNewData(uint32_t dt)
 }
 #endif
 
+void imuQuaternionComputeProducts(quaternion *quat) {
+    quat->ww = quat->w * quat->w;
+    quat->wx = quat->w * quat->x;
+    quat->wy = quat->w * quat->y;
+    quat->wz = quat->w * quat->z;
+    quat->xx = quat->x * quat->x;
+    quat->xy = quat->x * quat->y;
+    quat->xz = quat->x * quat->z;
+    quat->yy = quat->y * quat->y;
+    quat->yz = quat->y * quat->z;
+    quat->zz = quat->z * quat->z;
+}
+
+bool imuQuaternionHeadfreeDislocationSet(void) {
+    if((fabsf(attitude.values.roll/10.0f) < 45.0f)  && (fabsf(attitude.values.pitch/10.0f) < 45.0f)) {
+        const float roll = 0;
+        const float pitch = 0;
+        const float yaw = -atan2f((+2.0f * (q.wz + q.xy)), (+1.0f - 2.0f * (q.yy + q.zz)));
+
+        const float cosRoll2 = cos(roll/2);
+        const float cosPitch2 = cos(pitch/2);
+        const float cosYaw2 = cos(yaw/2);
+        const float sinRoll2 = sin(roll/2);
+        const float sinPitch2 = sin(pitch/2);
+        const float sinYaw2 = sin(yaw/2);
+        const float cosPitch2cosYaw2 = cosPitch2 * cosYaw2;
+        const float sinPitch2sinYaw2 = sinPitch2 * sinYaw2;
+
+        dislocation.w = cosRoll2 * cosPitch2cosYaw2 + sinRoll2 * sinPitch2sinYaw2 ;
```

Comment:
It can eliminate half of multiplications in `imuQuaternionMultiplication`

It will also allow modifying rotation matrix directly, saving 15 of 27 multiplications in matrix multiplication (geting same performance as generic quaternion multiplication) and avoiding quaternion-matrix conversion, (12 mul, 12 add) (which is effectively inlined in your code). No drift problem here, so there is little disadvantage except for constraining rotation to single axis.